pennylane/fourier/utils.py

```diff
@@ -42,3 +42,73 @@ def to_dict(coeffs):
     nvecs = product(*n_ranges)
 
     return {nvec: coeffs[nvec] for nvec in nvecs}
+
+
+def _get_spectrum(op, decimals):
+    r"""Extract the frequencies contributed by an input-encoding gate to the
+    overall Fourier representation of a quantum circuit.
+
+    If :math:`G` is the generator of the input-encoding gate :math:`\exp(-i x G)`,
+    the frequencies are the differences between any two of :math:`G`'s eigenvalues.
+    We only compute non-negative frequencies in this subroutine.
+
+    Args:
+        op (~pennylane.operation.Operation): :class:`~.pennylane.Operation` to extract
+            the frequencies for
+        decimals (int): Number of decimal places to round the frequencies to
+
+    Returns:
+        set[float]: non-negative frequencies contributed by this input-encoding gate
+    """
+    no_generator = False
+    if hasattr(op, "generator"):
+        g, coeff = op.generator
+
+        if isinstance(g, np.ndarray):
+            matrix = g
+        elif hasattr(g, "matrix"):
+            matrix = g.matrix
+        else:
+            no_generator = True
+    else:
+        no_generator = True
+
+    if no_generator:
+        raise ValueError(f"Generator of operation {op} is not defined.")
+
+    matrix = coeff * matrix
+    # todo: use qml.math.linalg once it is tested properly
+    evals = np.linalg.eigvalsh(matrix)
+
+    # compute all unique positive differences of eigenvalues, then add 0
+    # note that evals are sorted already
+    _spectrum = set(np.round([x[1] - x[0] for x in combinations(evals, 2)], decimals=decimals))
+    _spectrum |= {0}
```

Comment:
Here we can use that `evals` is sorted already, so that we know `x[1]-x[0]` to be non-negative. `0` is always included in a spectrum because it contributes to combinations when using `_join_spectra`